UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaTileMap.cs

```diff
@@ -1,587 +1,1316 @@
+				{
+					(layer as UnderFloorLayer).RemoveSpecifiedTile(position, null, true);
+				}
+
+
+				TileLocation TileLcation = null;
+				lock (PresentTiles)
+				{
+					PresentTiles[layer].TryGetValue(position, out TileLcation);
+				}
+
+				if (TileLcation != null)
+				{
+					TileLcation.Tile = null;
+					TileLcation.OnStateChange();
+				}
+
+				//layer.RemoveTile(position);
+			}
+			else
+			{
+				LogMissingLayer(position, refLayer);
+			}
 		}
-	}
 
-	public void ClearAllTiles()
-	{
-		for (var i = 0; i < LayersValues.Length; i++)
+		public void ClearAllTiles()
 		{
-			LayersValues[i].ClearAllTiles();
+			for (var i = 0; i < LayersValues.Length; i++)
+			{
+				lock (PresentTiles)
+				{
+					foreach (var keypar in PresentTiles[LayersValues[i]])
+					{
+						keypar.Value.Clean();
+						PooledTileLocation.Push(keypar.Value);
+					}
+
+					PresentTiles[LayersValues[i]].Clear();
+				}
+
+				LayersValues[i].ClearAllTiles();
+			}
 		}
-	}
 
-	public Vector3 LocalToWorld( Vector3 localPos ) => LayersValues[0].LocalToWorld( localPos );
-	public Vector3 CellToWorld( Vector3Int cellPos ) => LayersValues[0].CellToWorld( cellPos );
-	public Vector3 WorldToLocal( Vector3 worldPos ) => LayersValues[0].WorldToLocal( worldPos );
+		public Vector3 LocalToWorld(Vector3 localPos) => LayersValues[0].LocalToWorld(localPos);
+		public Vector3 CellToWorld(Vector3Int cellPos) => LayersValues[0].CellToWorld(cellPos);
+		public Vector3 WorldToLocal(Vector3 worldPos) => LayersValues[0].WorldToLocal(worldPos);
 
-	public BoundsInt GetWorldBounds()
-	{
-		var bounds = GetBounds();
-		//???
-		var min = CellToWorld(bounds.min);
-		var max = CellToWorld(bounds.max);
+		public BoundsInt GetWorldBounds()
+		{
+			var bounds = GetBounds();
+			//???
+			var min = CellToWorld(bounds.min);
+			var max = CellToWorld(bounds.max);
+			if (PresentMatrix?.MatrixMove?.inProgressRotation != null)
+			{
+				Vector3Int TopRightMax = bounds.max;
+				Vector3Int BottomLeftMin = bounds.min;
+				Vector3Int BottomRight = new Vector3Int(TopRightMax.x, BottomLeftMin.y, 0);
+				Vector3Int TopLeft = new Vector3Int(BottomLeftMin.x, TopRightMax.y, 0);
+				var TopRightMaxI = CellToWorld(TopRightMax);
+				var BottomLeftMinI = CellToWorld(BottomLeftMin);
+				var BottomRightI = CellToWorld(BottomRight);
+				var TopLeftI = CellToWorld(TopLeft);
+				MaxMinCheck(ref min, ref max, TopRightMaxI);
+				MaxMinCheck(ref min, ref max, BottomLeftMinI);
+				MaxMinCheck(ref min, ref max, BottomRightI);
+				MaxMinCheck(ref min, ref max, TopLeftI);
+			}
 
-		return new BoundsInt(min.RoundToInt(), (max - min).RoundToInt());
-	}
 
-	public BoundsInt GetBounds()
-	{
-		Vector3Int minPosition = Vector3Int.one * int.MaxValue;
-		Vector3Int maxPosition = Vector3Int.one * int.MinValue;
+			return new BoundsInt(min.RoundToInt(), (max - min).RoundToInt());
+		}
 
-		for (var i = 0; i < LayersValues.Length; i++)
+		public void MaxMinCheck(ref Vector3 min, ref Vector3 max, Vector3 ToCompare)
```

Comment:
Codacy found an issue: [Consider refactoring this method in order to remove the need for this 'ref' modifier.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6319358)